### PR TITLE
EdgeHub to module messages: Make module response timeout configurable

### PR DESCRIPTION
**Pull request by and-rewsmith**
This timeout can now be configured with a new environment variable in the edge hub deployment:
```MessageAckTimeoutSecs```

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs [hunk 1]
```diff
@@ -17,22 +17,25 @@ namespace Microsoft.Azure.Devices.Edge.Hub.Core.Device
...
         IDeviceProxy underlyingProxy;
 
-        public DeviceMessageHandler(IIdentity identity, IEdgeHub edgeHub, IConnectionManager connectionManager)
+        public DeviceMessageHandler(IIdentity identity, IEdgeHub edgeHub, IConnectionManager connectionManager, TimeSpan messageResponseTimeout)
         {
             this.Identity = Preconditions.CheckNotNull(identity, nameof(identity));
             this.edgeHub = Preconditions.CheckNotNull(edgeHub, nameof(edgeHub));
             this.connectionManager = Preconditions.CheckNotNull(connectionManager, nameof(connectionManager));
+            this.messageResponseTimeout = Preconditions.CheckNotNull(messageResponseTimeout, nameof(messageResponseTimeout));
+            Events.Log.LogDebug($"Message response timeout: {this.messageResponseTimeout.ToString()}");
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs]: Log this once in the beginning instead

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs [hunk 2]
```diff
@@ -251,7 +254,7 @@ async Task HandleTwinOperationException(string correlationId, Exception e)
         static class Events
         {
             const int IdStart = HubCoreEventIds.DeviceListener;
-            static readonly ILogger Log = Logger.Factory.CreateLogger<DeviceMessageHandler>();
+            public static readonly ILogger Log = Logger.Factory.CreateLogger<DeviceMessageHandler>();
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs]: Don't do this.. add a method to this class to log the event.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/DependencyManager.cs [hunk 3]
```diff
@@ -135,6 +135,8 @@ void RegisterMqttModule(ContainerBuilder builder, (bool isEnabled, bool usePersi
             TimeSpan connectivityCheckFrequency = connectivityCheckFrequencySecs < 0 ? TimeSpan.MaxValue : TimeSpan.FromSeconds(connectivityCheckFrequencySecs);
             // n Clients + 1 Edgehub
             int maxConnectedClients = this.configuration.GetValue("MaxConnectedClients", 100) + 1;
+            int messageResponseTimeoutSecs = this.configuration.GetValue("MessageResponseTimeoutSecs", 30);
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/DependencyManager.cs]: Suggest calling this env var - MessageAckTimeoutSecs

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs (hunk 1 above)
Comment (and-rewsmith) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs]: This was just for debugging. See commit https://github.com/Azure/iotedge/pull/2652/commits/d596b1889f93a6197c8eb190db5e47559f6f8d87

Comment (and-rewsmith) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs]: Will remove after I verify. Will also log in the beginning.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs [hunk 4]
```diff
@@ -66,6 +66,7 @@ static async Task<int> MainAsync(IConfigurationRoot configuration)
             LogLogo(logger);
             LogVersionInfo(logger);
             logger.LogInformation($"OptimizeForPerformance={configuration.GetValue("OptimizeForPerformance", true)}");
+            logger.LogInformation($"MessageAckTimeoutSecs={configuration.GetValue("MessageAckTimeoutSecs", 30)}");
```
Comment (and-rewsmith) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs]: I followed the same pattern for OptimizeForPerformance, but this seems error prone as we specify the defaults here as well as DependencyManager.cs. We should explore cleaning this up somehow in the future.

File: edge-hub/test/Microsoft.Azure.Devices.Edge.Hub.Core.Test/DeviceMessageHandlerTest.cs [hunk 5]
```diff
@@ -200,27 +202,67 @@ public async Task MessageRejectedTest()
...
+
+            string messageId = receivedMessage.SystemProperties[SystemProperties.LockToken];
+            await deviceMessageHandler.ProcessMessageFeedbackAsync(messageId, FeedbackStatus.Complete);
+
+            await Task.Delay(TimeSpan.FromSeconds(1));
             await Assert.ThrowsAsync<TimeoutException>(async () => await sendMessageTask);
         }
 
+        [Fact]
+        public async Task MessageCompletionLongAckTimeoutTest()
```
Comment (and-rewsmith) [inline edge-hub/test/Microsoft.Azure.Devices.Edge.Hub.Core.Test/DeviceMessageHandlerTest.cs]: Added two tests:
MessageCompletionShortAckTimeoutTest: A short ack timeout will correctly timeout when the device/module takes a longer time to ack.
MessageCompletionLongAckTimeoutTest: A long ack timeout will correctly not timeout when the device/module takes a shorter time to ack.

**Pull request merged**